### ISSUE TITLE
Booleans as bits/`uint1`s?

### ISSUE THREAD

**Issue by alexandergunnarson** [DrTimothyAldenDavis/SuiteSparse#219]
**Is your feature request related to a problem? Please describe.**
This is an idea around enhancing the performance (both space and time) of boolean tensors. (The optimization proposed may already be implemented in SuiteSparse, or it may have been explored and found not to be worth it in practice. Alternatively, it may work well on GPU, but currently not CPU.) We're exploring using SuiteSparse's implementation of GraphBLAS for a problem we have at my startup, and we'd like to ensure it handles a certain level of data scale we need to support.

**Describe the solution you'd like**
The proposed solution is to implement booleans in GraphBLAS with the smallest datatype possible rather than as e.g. straightforward `uint32`s/`uint8`s, etc., in order to save on space and exploit maximal hardware parallelism. This might be directly `uint1`s on GPU (supported in CUDA), or bits within raw `uint8`s on CPU (manipulated with bitwise operations).

**Describe alternatives you've considered**
The alternative is to use a "straightforward" (e.g. `uint32`/`uint8` as numeric `0`/`1`) implementation of boolean values.

**Additional context**
In theory, if we measure baseline performance on, say, booleans as `uint8`s, we might have as much as 8x better performance with booleans as `uint1`s. (Of course, the speedup might not get this far in practice, because on CPU you'd ostensibly need to work with vectorized bit ops, which likely perform differently than vectorized addition/multiplication.)

Motivating papers:
- [SPbLA: The Library of GPGPU-powered Sparse Boolean Linear Algebra Operations (8/2022)](https://www.researchgate.net/publication/362825622_SPbLA_The_Library_of_GPGPU-powered_Sparse_Boolean_Linear_Algebra_Operations) (repo available at https://github.com/SparseLinearAlgebra/spbla)
- Relatedly, [cuBool: Bit-Parallel Boolean Matrix Factorization on CUDA-Enabled Accelerators (2018)](https://ieeexplore.ieee.org/abstract/document/8644574) (repo available at https://github.com/SparseLinearAlgebra/cuBool)

Motivating benchmark results (2021):
![image](https://user-images.githubusercontent.com/4691106/207519923-02b5f6da-795c-436c-bbc4-f6afee546d28.png)
![image](https://user-images.githubusercontent.com/4691106/207519936-309acba0-06ed-4c90-a05f-674eeae4e6c8.png)

**Comment by DrTimothyAldenDavis** [comment]
I've thought about using single bits for boolean types, but I haven't taken the plunge.  It complicates the parallelism of the method quite a bit, since multiple threads can't write individual bits without complicated atomics or unless the work is partitioned so that any one thread accesses all 8 bits in the uint8.  Either option is difficult and it would change my algorithms quite a bit.

The most common use of booleans I've seen so far, in LAGraph, is a placeholder for 'no value needed at all'.  The matrix becomes just structural.  I can now do this for any matrix type: bool, int*, float, double, etc, for what I call 'iso-value' matrices.  If all entries of a matrix have the same value (say, 1, or 3.14, or whatever), then I just store it once.

With iso-value matrices, there's somewhat less need for boolean matrix multiply, at least in the LAGraph algorithms we have so far.

The 2nd paper is on boolean matrix factorization, and for that, you can't use the iso-value trick, so storing the values as individual bits would more important there.

The first paper you mention on SpBla doesn't say which version of SuiteSparse:GraphBLAS was used.  They cite my paper with the single-threaded version from 2019, but I would hope they'd be using a more recent parallel version.  It's not clear at all which they used.  That's a pity since it would have a big impact on their benchmarks.  I'd still expect the GPU-based methods to be faster than SuiteSparse on the CPU, of course.

**Comment by alexandergunnarson** [comment]
I was envisioning that kind of complication on the CPU, yeah. I wonder, however, if execution on the GPU changes things. On the CPU, the smallest addressable datatype is (generally) `uint8` of course, but from my extremely limited knowledge of CUDA (https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#built-in-vector-types), `uint1` is supported as a datatype. Perhaps it's as simple as changing a CUDA kernel from `uint8` to `uint1` (granted, keeping in mind the JIT functionality you mentioned in a different conversation). But I'm sure the reality is more complicated — there are GPU cache characteristics and perhaps other hardware considerations that might make it so that `uint1` ends up counterintuitively being less efficient than something more widely supported on GPU like, say, `uint4`.

Very interesting about iso-valued matrices! I checked out your paper that discusses it further (["Algorithm 10xx: SuiteSparse:GraphBLAS: parallel graph algorithms in the language of sparse linear algebra" (1/2022)](https://raw.githubusercontent.com/DrTimothyAldenDavis/GraphBLAS/stable/Doc/toms_parallel_grb2.pdf)). Curious how there's somewhat less need for boolean matrix multiply with iso-valued matrices. Could you expand a bit? Perhaps this is where I'll need to read rather than skim your paper I linked.

As for the SPbLA paper — that's a good point that I also considered. I'd have to assume it's using the parallel version of SuiteSparse:GraphBLAS, or else it's really not a fair or even useful comparison. (And even then, it'd be important to see how many CPU cores were engaged, which they didn't report.) Ostensibly it's using v6.1.4, per [the benchmark repo](https://github.com/EgorOrachyov/graph-bench/blob/85afc5ea93041e3b5b5f3341be3fe723b97763aa/scripts/build_lagraph.py#L28) they used, but it's still unclear, as you say. In any case the comparison to cuSPARSE seems more illustrative, and however inadequate their benchmarking methods may have been, their boolean-as-bit approach seemed particularly worth mining for ideas.

**Comment by DrTimothyAldenDavis** [comment]
Thanks for pointing out the benchmark repo link (I hadn't found that).  GraphBLAS v6.1.4 is probably fine for their benchmarks.  It was released on Jan 13, 2022, and they did their benchmarks in Feb 2022, so that's a recent version for them.  I've made a few changes since then to speed up the sparse-times-sparse matrix multiply, but I think it's likely that the changes I've made since then don't have much effect on their benchmarks.

The only question remains is:  did they have iso-valued matrices, as is often the case for boolean matrices?  That would have a big impact on the results.

The prevalence of iso-valued problems comes from my experience in LAGraph.  It's quite often that input graphs are iso-valued (representing unweighted graphs, which are very common).  So in that case, a LOR_LAND_BOOL semiring would see input matrices A and B with all entries in the 2 matrices (that are present) equal to TRUE.  The C matrix for C=A*B would have the same property.  I exploit this fact by computing just one value for C (true), and so I only have to compute the pattern.  Essentially, I use the "ANY_PAIR_BOOL" semiring, or "ANY_ONEB_BOOL" as it's now called, if I detect the case where C is iso-valued.  That semiring just computes the pattern of C.

See: https://github.com/DrTimothyAldenDavis/GraphBLAS/blob/f1471c711d87263afe43df55d586c442059c0361/Source/GB_AxB_saxpy3.c#L625-L639

And I determine the iso property of C based on the semiring and the inputs A and B:
https://github.com/DrTimothyAldenDavis/GraphBLAS/blob/stable/Source/GB_iso_AxB.c

There are many other cases where a boolean matrix would be non-iso-valued, such as in a boolean matrix factorization.
